src/server/graph-generation/treeTraversal.js

```diff
@@ -0,0 +1,114 @@
+/**
+    Pathway Commons Central Data Cache
+
+    Tree Traversal
+    treeTraversal.js
+
+    Purpose  : To provide functions to search and traverse a tree represented using 
+               javascript arrays
+
+    Requires : Valid tree array
+
+    @author Harsh Mistry
+    @version 1.1 2017/10/17
+**/
+
+//Returns all values mapped to a key in a given subtree
+//Requires subtree to be valid
+//Note : - A empty array is returned if no match is found
+//       - Recurse indicates if other levels of the tree should be searched
+function searchTree(subtree, key, recurse = true) {
+    var result = []
+  
+    //Loop through all level nodes
+    for (var i = 0; i < subtree.length; i++) {
+      //Push current value if it matches
+      if (subtree[i][0] == key) {
+        result.push(subtree[i][1]);
+      }
+  
+      //Recurse on subtree if one exists
+      if (subtree[i][1] instanceof Array && recurse) {
+        result.push(searchTree(subtree[i][1], key));
+      }
+    }
+  
+    //Remove all invalid values
+    for (var i = 0; i < result.length; i++) {
+      if (!(result[i])) {
+        result.splice(i, 1);
+      }
+    }
+  
+    return result;
+  }
+  
+  //Search for just one entry
+  //Requires a valid subtree
+  //Note : Recurse will search entire subtree for the node
+  function searchOne(subtree, key, name) {
+    var temp = searchTree(subtree, key, false);
+    if (temp.length > 0) return name + temp[0];
+    else return null;
+  }
+  
+  //Search for multiple entries
+  //Requires a valid subtree
+  function searchMultiple(subtree, key, name) {
+    var temp = searchTree(subtree, key, false);
+    if (temp && temp.length > 0) return [name, temp];
+    else return null;
+  }
+  
+  
+  //Search for a subnode
+  //Returns an array, a string, or null
+  //Requires a valid subtree
+  function searchForNode(subtree, key) {
+    if (!(subtree)) return null;
+    for (var i = 0; i < subtree.length; i++) {
+      if (subtree[i][0].indexOf(key) !== -1) return subtree[i][1];
```

Comment:
If you search for tag `<bar>` and see `<foobar>`, this code would also return the `<foobar>` tags (i.e. `'bar'` is a substring of `'foobar'`)

Reply:
Parser